src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1762,31 +1900,106 @@ public sealed class One : Tests
                 public override void Filter(
                     DecisionDagBuilder builder,
                     BoundDagTest test,
+                    DagState state,
                     IValueSet? whenTrueValues,
                     IValueSet? whenFalseValues,
                     out Tests whenTrue,
                     out Tests whenFalse,
                     ref bool foundExplicitNullTest)
                 {
+                    if (Test is BoundDagEvaluation)
+                    {
+                        // there cannot be any implication for an evaluation node.
+                        whenTrue = whenFalse = this;
+                        return;
+                    }
+
                     builder.CheckConsistentDecision(
                         test: test,
                         other: Test,
+                        state: state,
                         whenTrueValues: whenTrueValues,
                         whenFalseValues: whenFalseValues,
                         syntax: test.Syntax,
+                        precondition: out Tests precondition,
+                        sideeffect: out Tests sideeffect,
                         trueTestPermitsTrueOther: out bool trueDecisionPermitsTrueOther,
                         falseTestPermitsTrueOther: out bool falseDecisionPermitsTrueOther,
                         trueTestImpliesTrueOther: out bool trueDecisionImpliesTrueOther,
                         falseTestImpliesTrueOther: out bool falseDecisionImpliesTrueOther,
                         foundExplicitNullTest: ref foundExplicitNullTest);
-                    whenTrue = trueDecisionImpliesTrueOther ? Tests.True.Instance : trueDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
-                    whenFalse = falseDecisionImpliesTrueOther ? Tests.True.Instance : falseDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
+                    // Given T0 as a test that has already occurred, T1 as a subsequent test, P as a pre-condition, and S as a side-effect (expected to always evaluate to True), we proceed as follow:
+                    //  - If T0=True implies T1, we rewrite T1 as (P ? S : T1) because we have determined that given T0=True, T1 would always succeed if P evaluates to True.
+                    //    Note: If there is no pre-condition, i.e. P is True, the above will be further simplified as True which means T1 is insignificant.
+                    //  - If T0=True disallows T1, we rewrite T1 as (P ? !S : T1) because we have determined that given T0=True, T1 would never succeed if P evaluates to True.
+                    //    Note: If there is no pre-condition, i.e. P is True, the above will be further simplified as False which means T1 is impossible.
+                    //  - Otherwise, we rewrite T1 as ((!P || S) && T1) to preserve the side-effect if the pre-condition holds.
```

Comment:
nit: let's swap bullets 2 (If T0=True disallows T1) and 3 (If T0=True allows T1) to match code order below

Reply:
> [If T0=True allows T1], we rewrite T1 as ((!P || S) && T1) to preserve the side-effect if the pre-condition holds.

I didn't follow that. Why aren't we just using `T1` in this case?
T0 and T1 are unrelated, even when the precondition applies, so we must evaluate T1 (which gives us remaining values) and there is no other information we can use to factor into those remaining values.

Reply:
>T0 and T1 are unrelated

They may be related but with no implications (which is where we pick this codepath). 
I think this matters for the optimization I mentioned above https://github.com/dotnet/roslyn/pull/53891#discussion_r691470124

Reply:
My disturbing lack of tdd discipline shows :smile: The rewrite is required as it affect some of tests.
(Will revert the prototype comment in the next iteration.)

Reply:
It took me a while, but I think I finally get the third case. Pinged Fred to double check.

"They may be related but with no implications from one to the other (which is where we pick this codepath)."
I was mixing "unrelated" and "has no implication". In the case where "test has no implication on other" it's still possible that the two are "related" (ie. they turn out to test the same thing when `precondition` is true). In such case, we need `sideeffect` to track the merging of the remaining values.